src/main/java/org/terasology/web/webSocket/ClientMessage.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.web.webSocket;
+
+import com.google.gson.JsonElement;
+
+public class ClientMessage {
+
+    public enum MessageType {
+        AUTHENTICATION_REQUEST,
+        AUTHENTICATION_DATA,
+        RESOURCE
+    }
+
+    private MessageType messageType;
+    private JsonElement data;
+
+    public MessageType getMessageType() {
+        return messageType;
+    }
+
+    public JsonElement getData() {
+        return data;
+    }
+
+    public void checkValid() throws InvalidClientMessageException {
```

Comment:
I agree that relying on the caller to perform checks isn't very good, however at the moment I can't find any real alternative. I read the InstanceCreator documentation and as far as I understand it allows you to define code that will run **before** the deserialization and must return an instance, which fields will then be overwritten with the deserialized data. So, in the [T createInstance(Type type)](https://static.javadoc.io/com.google.code.gson/gson/2.6.2/com/google/gson/InstanceCreator.html#createInstance-java.lang.reflect.Type-) method I don't have access to the data and thus can't validate it.

I also considered using a custom [deserializer](https://static.javadoc.io/com.google.code.gson/gson/2.6.2/com/google/gson/JsonDeserializer.html), but I'm not sure it would be of much help:
- I should find a way of deserializing the data anyway; using context.deserialize() would cause an infinite loop because it would call the same `deserialize` method, so the remaining options are manual deserialization (probably not that hard, but would defeat the purpose of using gson I guess?) or creating a new GSON instance (losing the context, i.e. registered adapters; in this case it shouldn't be a big deal since there are just a `String` field and a `JsonElement` one, however the other class where I should perform similar checks to avoid the NullReferenceException catch block we are talking about in [the other comment thread](https://github.com/MovingBlocks/FacadeServer/pull/2#discussion_r122599388) - `ClientAuthenticationMessage` - would be much more troubling because the structure contains `BigInteger` and `byte[]` fields, which have custom deserializers).
- I couldn't throw a custom exception such as `InvalidClientMessageException` because of the signature of the method in the interface that only allows to throw `JsonParseException`. This probably isn't much of a problem though since I could throw it as the cause of the JsonParseException.